### PR TITLE
Hauling overhaul

### PR DESCRIPTION
**Pull request by actually-a-cat**
#### Summary
Features "Hauling overhaul: you can choose which specific items to haul, whether to automatically haul new items, or haul items by filter"

#### Purpose of change
Hauling picks up every item stepped on, which causes some usability issues. If you want to haul only some specific items, you may need to navigate around anything on the floor, or stop to untangle piles. When used unattended (e.g. while autotravelling) there is risk that it will pick up junk along the way and slow you down unnecessarily. 

#### Describe the solution
image

Pressing the haul button now brings up a little UI where items to be hauled can be chosen, automatically including new items to haul can be enabled/disabled and a filter for the automatic haul can be set. An action to replicate old behavior (not bound by default) is available. Start/stop hauling is always hotkeyed to the same button as the key to open the menu, so a quick double-tap can be used for the most common action.

Caveats:
* this feature does not support `count_by_charges`. It will haul those safely, but it's not possible to haul part of a charge stack, and any stacks of a charge item already hauled will be merged into the haul even with autohaul off. I'm choosing to leave it like this weighing this minor inconvenience against the significant effort of supporting charges in light of them being a feature on the way out.
* hauling will lose track of items that spill to other tiles due to the destination tile being full. They are handled safely but it's up to the player to sort out the mess; there is a warning message in the log when it happens.

#### Testing
- [x] basic hauling of a handful of items
- [x] hauling one stack over another with autohaul off does not mix them up
- [x] hauling one stack over another with autohaul on gathers them together
- [x] partially combine stacks with autohaul and filter
- [x] choose part of a stack with the selector
- [x] combine chosen part of a stack with another stack, confirm selector state makes sense
- [x] overflowing a tile produces a warning and doesn't lose any items

### CODE REVIEW

**Comment by Zephanos** [comment]
Having a toggle to only haul what what was on the initial tile the button was pressed would also be a good option. That way when your looting an area you can quickly haul a specific group of items you may already have sorted without messing with the filter. Likewise, if you do have a filter set, once you gather everything you want you can turn on this option and move freely without worry of picking up other random junk that might also be in the set filter.

**Comment by actually-a-cat** [comment]
> Having a toggle to only haul what was on the initial tile the button was pressed would also be a good option. 

This is what happens if you enable filtering and leave the filter empty. It does need to be more obvious how it works, I'll keep thinking about the UI.

File: src/handle_action.cpp [hunk 1]
```diff
@@ -730,21 +730,85 @@ static void haul()
...
+    bool exit = false;
+    do {
+        uilist menu;
+
+        bool hauling = player_character.is_hauling();
+        bool selective = player_character.is_hauling_selectively();
+
+        std::string header = _( "You are currently " );
+
+        header += hauling ? _( "hauling items." ) : _( "not hauling items." );
```
Comment (Qrox) [inline src/handle_action.cpp:742]:
```suggestion
        std::string header = hauling ? _( "You are currently hauling items." ) : _( "You are currently not hauling items." );
```
Other languages do not necessarily have the same word order as English.

**Comment by Qrox** [comment]
It would also be nice if you can select what specific items to haul with an inventory UI.

**Pull request merged**